contracts/Committee.sol

```diff
@@ -5,79 +5,224 @@ import "./spec_interfaces/IValidatorsRegistration.sol";
-		uint32 readyToSyncTimestamp;
-		bool isMember; // exists
-		bool readyForCommittee;
+		uint96 weight;
+		uint8 pos;
+		bool isMember;
 		bool isCompliant;
 
-		uint8 role;
+		bool inCommittee;
 	}
 	mapping (address => MemberData) membersData;
 
-	struct Member {
-		address addr;
-		MemberData data;
-	} // Never in state, only in memory
-
-	struct Participant {
-		MemberData data;
-		address addr;
-		uint8 pos;
-
-		uint8 oldRole;
-		uint8 newRole;
-	} // Never in state, only in memory
-
-	struct Settings { // TODO can be reduced to 2-3 state entries
-		uint32 readyToSyncTimeout;
-		uint32 maxTimeBetweenRewardAssignments;
-		uint8 maxCommitteeSize;
-		uint8 maxStandbys;
-	}
-	Settings settings;
-
-	uint256 weightSortIndicesOneBasedBytes;
-
 	// Derived properties
 	struct CommitteeInfo {
-		uint64 committeeBitmap;
-		uint8 standbysCount;
+		uint32 committeeBitmap; // TODO redundant, sort bytes can be used instead
+		uint8 minCommitteeMemberPos;
 		uint8 committeeSize;
 	}
 	CommitteeInfo committeeInfo;
+	bytes32 committeeSortBytes;
+
+	struct Settings {
+		uint32 maxTimeBetweenRewardAssignments;
+		uint8 maxCommitteeSize;
+	}
+	Settings settings;
 
 	modifier onlyElectionsContract() {
 		require(msg.sender == address(getElectionsContract()), "caller is not the elections");
 
 		_;
 	}
 
-	constructor(uint _maxCommitteeSize, uint _maxStandbys, uint256 _readyToSyncTimeout, uint32 maxTimeBetweenRewardAssignments) public {
+	function findFreePos(CommitteeInfo memory info) private pure returns (uint8 pos) {
+		pos = 0;
+		uint32 bitmap = info.committeeBitmap;
+		while (bitmap & 1 == 1) {
+			bitmap >>= 1;
+			pos++;
+		}
+	}
+
+	function qualifiesToEnterCommittee(address addr, MemberData memory data, CommitteeInfo memory info, Settings memory _settings) private view returns (bool qualified, uint8 entryPos) {
+		if (!data.isMember || data.weight == 0) {
+			return (false, 0);
+		}
+
+		if (info.committeeSize < _settings.maxCommitteeSize) {
+			return (true, findFreePos(info));
+		}
+
+		CommitteeMember memory minMember = committee[info.minCommitteeMemberPos];
+		if (data.weight < minMember.weight || data.weight == minMember.weight && addr < minMember.addr) {
+			return (false, 0);
+		}
+
+		return (true, info.minCommitteeMemberPos);
+	}
+
+	function saveMemberData(address addr, MemberData memory data) private {
+		if (data.isMember) {
+			membersData[addr] = data;
+		} else {
+			delete membersData[addr];
+		}
+	}
+
+	function updateOnMemberChange(address addr, MemberData memory data) private returns (bool committeeChanged) {
+		CommitteeInfo memory info = committeeInfo;
+		Settings memory _settings = settings;
+		bytes memory sortBytes = abi.encodePacked(committeeSortBytes);
+
+		if (!data.inCommittee) {
+			(bool qualified, uint8 entryPos) = qualifiesToEnterCommittee(addr, data, info, _settings);
+			if (!qualified) {
+				saveMemberData(addr, data);
+				return false;
+			}
+
+			(info, sortBytes) = removeMemberAtPos(entryPos, sortBytes, info);
```

Comment:
It's a bit confusing that removeMemberAtPos updates the members map and otherwise its explicitly set by saveMemberData. But i currently don't have a nicer option in mind.

Reply:
Right, would be nicer to use `saveMemberData` there also. Will change